File: android/src/main/java/com/ultralytics/ultralytics_yolo/ResultStreamHandler.java
package com.ultralytics.ultralytics_yolo;

import android.os.Handler;
import android.os.Looper;

import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.EventChannel;

class ResultStreamHandler implements EventChannel.StreamHandler {
    final private Handler handler = new Handler(Looper.getMainLooper());
    private EventChannel.EventSink eventSink;

    @Override
    public void onListen(Object arguments, EventChannel.EventSink events) {
        eventSink = events;
    }

    @Override
    public void onCancel(Object arguments) {
        eventSink = null;
    }

    public void sink(List<Map<String, Object>> objects) {
        handler.post(() -> {
            if (eventSink != null && !objects.isEmpty()) {
                eventSink.success(objects);
            }
        });
    }

    public void close() {
        if (eventSink != null) {
            eventSink.endOfStream();
            eventSink = null;
        }
    }
}

File: ios/Classes/ResultStreamHandler.swift
//
//  ResultStreamHandler.swift
//  ultralytics_yolo
//
//  Created by Sergio Sánchez on 9/11/23.
//

import Foundation

class ResultStreamHandler: NSObject, FlutterStreamHandler {
    private let handler = DispatchQueue.main
    private var eventSink: FlutterEventSink?

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
    
    func sink(objects: [[String: Any]]) {
        handler.async {
            if let eventSink = self.eventSink, !objects.isEmpty {
                eventSink(objects)
            }
        }
    }
    
    func close() {
        if let eventSink = eventSink {
            eventSink(FlutterEndOfEventStream)
            self.eventSink = nil
        }
    }
}
